crates/iota-data-ingestion-core/src/reader/v2.rs

```diff
@@ -0,0 +1,717 @@
+            self.last_pruned_watermark,
+            checkpoints.len(),
+        );
+        for checkpoint in checkpoints {
+            assert_eq!(
+                checkpoint.checkpoint_summary.sequence_number,
+                self.current_checkpoint_number
+            );
+            self.checkpoint_tx.send(checkpoint).await.map_err(|_| {
+                IngestionError::Channel(
+                    "unable to send checkpoint to executor, receiver half closed".to_owned(),
+                )
+            })?;
+            self.current_checkpoint_number += 1;
+        }
+
+        Ok(())
+    }
+
+    async fn run(mut self) {
+        let (_watcher, mut inotify_rx) = setup_directory_watcher(&self.path);
+
+        gc_processed_files(
+            &self.path,
+            self.last_pruned_watermark,
+            &mut self.last_pruned_watermark,
+            &mut self.data_limiter,
+        )
+        .expect("Failed to clean the directory");
+
+        loop {
+            tokio::select! {
+                _ = &mut self.shutdown_rx => break,
+                Some(watermark) = self.gc_signal_rx.recv() => {
+                    gc_processed_files(
+                        &self.path,
+                        watermark,
+                        &mut self.last_pruned_watermark,
+                        &mut self.data_limiter,
+                    ).expect("Failed to clean the directory");
+                }
+                Ok(Some(_)) | Err(_) = timeout(Duration::from_millis(self.tick_interval_ms), inotify_rx.recv())  => {
+                    self.sync().await.expect("Failed to read checkpoint files");
+                }
+            }
+        }
+    }
+}
+
+/// Public API for interacting with the checkpoint reader actor.
+///
+/// It provides methods to receive streamed checkpoints, send garbage collection
+/// signals, and gracefully shut down the background checkpoint reading task.
+/// Internally, it communicates with a [`CheckpointReaderActor`], which manages
+/// the actual checkpoint fetching and streaming logic.
+pub(crate) struct CheckpointReader {
+    handle: JoinHandle<()>,
+    shutdown_tx: oneshot::Sender<()>,
+    gc_signal_tx: mpsc::Sender<CheckpointSequenceNumber>,
+    checkpoint_rx: mpsc::Receiver<Arc<CheckpointData>>,
+}
+
+impl CheckpointReader {
+    pub(crate) async fn new(
+        starting_checkpoint_number: CheckpointSequenceNumber,
+        config: CheckpointReaderConfig,
+    ) -> IngestionResult<Self> {
+        let (checkpoint_tx, checkpoint_rx) = mpsc::channel(MAX_CHECKPOINTS_IN_PROGRESS);
+        let (gc_signal_tx, gc_signal_rx) = mpsc::channel(MAX_CHECKPOINTS_IN_PROGRESS);
+        let (shutdown_tx, shutdown_rx) = oneshot::channel();
+
+        let remote_fetcher: OptionFuture<_> = config
+            .remote_store_url
+            .map(|url| RemoteCheckpointFetcher::new(url, config.timeout_secs, config.batch_size))
+            .into();
+
+        let remote_fetcher = remote_fetcher.await.transpose()?;
```

Comment:
Nit: Explicit is better than implicit
```suggestion
        let remote_fetcher = if let Some(url) = config.remote_store_url {
            Some(RemoteCheckpointFetcher::new(url, config.timeout_secs, config.batch_size).await?)
        } else {
            None
        };
```

Reply:
fixed: https://github.com/iotaledger/iota/pull/7425/commits/5241e16fc87e6e1997c7325af2a426a5a988324d